### PR TITLE
Add state transitions to user

### PR DESCRIPTION
**Pull request by venkatd**

### CODE REVIEW

File: lib/code_corps/onboarding/user_transition.ex [hunk 1]
```diff
@@ -0,0 +1,10 @@
+defmodule CodeCorps.Onboarding.UserTransition do
```
Comment (joshsmith) [inline lib/code_corps/onboarding/user_transition.ex]: Wonder if `CodeCorps.Transition.UserState` is a better name?

Then we can also have a `Transition.PostStatus` or `PostStatus`.

File: web/models/user.ex [hunk 2]
```diff
@@ -28,6 +28,9 @@ defmodule CodeCorps.User do
     field :username, :string
     field :website, :string
 
+    field :state, :string
```
Comment (joshsmith) [inline web/models/user.ex:31]: Maybe add a `default` of `signed_up` here.

File: web/models/user.ex [hunk 3]
```diff
@@ -78,11 +81,22 @@ defmodule CodeCorps.User do
   def update_changeset(struct, params) do
     struct
     |> changeset(params)
-    |> cast(params, [:first_name, :last_name, :twitter, :biography, :website, :base64_photo_data])
+    |> cast(params, [:first_name, :last_name, :twitter, :biography, :website, :base64_photo_data, :state_transition])
```
Comment (joshsmith) [inline web/models/user.ex:84]: Think maybe we should `validate_inclusion` on the `state_transition` so the developer can track down the error. Even though `apply_state_transition()` does add one, that might be worthwhile to do.

File: web/models/user.ex (hunk 2 above)
Comment (joshsmith) [inline web/models/user.ex:31]: Oh I see that was there in the migration. Idk if you put that in `field`. My bad.

File: lib/code_corps/onboarding/user_transition.ex [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+  def next(current, nil), do: nil
+  def next("signed_up", "edit_profile"), do: {:ok, "edited_profile"}
+  def next("edited_profile", "select_categories"), do: {:ok, "selected_categories"}
+  def next("selected_categories", "select_roles"), do: {:ok, "selected_roles"}
+  def next("selected_roles", "select_skills"), do: {:ok, "selected_skills"}
+  def next(current, invalid_transition), do: {:error, "invalid transition #{invalid_transition} from #{current}"}
```
Comment (joshsmith) [inline lib/code_corps/onboarding/user_transition.ex]: Do we have a decent way to unit test these?

File: test/controllers/user_controller_test.exs [hunk 5]
```diff
@@ -207,6 +207,35 @@ defmodule CodeCorps.UserControllerTest do
...
+      conn = put authenticate(conn, user), user_path(conn, :update, user), %{
+        "data" => %{
+          "type" => "user",
+          "id" => user.id,
+          "attributes" => %{"password" => "password", "state_transition" => "yehaww"}
+        }
+      }
+
+      %{"errors" => errors} = json_response(conn, 422)
+      assert errors["state_transition"] == ["invalid transition yehaww from signed_up"]
```
Comment (joshsmith) [inline test/controllers/user_controller_test.exs:237]: You always did have the best fake data.

File: lib/code_corps/onboarding/user_transition.ex (hunk 1 above)
Comment (venkatd) [reply lib/code_corps/onboarding/user_transition.ex]: Yeah that's a better name.

File: lib/code_corps/onboarding/user_transition.ex (hunk 4 above)
Comment (venkatd) [reply lib/code_corps/onboarding/user_transition.ex]: It should be as simple as testing the return value of `next` a bunch of times. I personally don't think there's a lot of value in these sorts of tests (code is very simple and you have to maintain an additional test file). If someone else wants to take them on at a later time I can create an issue for it.

File: test/controllers/user_controller_test.exs (hunk 5 above)
Comment (venkatd) [reply test/controllers/user_controller_test.exs:237]: This line took me the longest to write.

File: web/models/user.ex (hunk 3 above)
Comment (venkatd) [reply web/models/user.ex:84]: The check seems redundant to me but I went ahead and added it anyway.

What's your reasoning behind it? If someone gets the transition name wrong, won't they go to the transition file?

Comment (joshsmith) [reply web/models/user.ex:84]: I mean if an Ember dev gets it wrong, just seems a little more explicit maybe?

Comment (joshsmith) [reply web/models/user.ex:84]: Wait, I missed that the error is added to `state_transition`, so this _is_ redundant. Sorry about that.

**Pull request merged**